2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+that will be assumed and discovered at runtime.
+Support for local filesystem repositories must be maintained.
+A `-InstallationPolicy` switch accepts `Trusted` or `Untrusted` (default) indicating
+whether to prompt the user when installing resources from that repository.
+
+`Get-PSResourceRepository` will list out the registered repositories.
+
+`Set-PSResourceRepository` can be used to update a repository URL or installation policy.
+
+`Unregister-PSResourceRepository` can be used to un-register a repository.
+
+### Finding resources
+
+Current design of PowerShellGet is to have different cmdlets for each of the
+different resources types that it supports:
+
+`Find-Command`
+`Find-DscResource`
+`Find-Module`
+`Find-RoleCapability`
+`Find-Script`
+
+Instead, these will be abstracted by a `Find-PSResource` cmdlet.
+A `-Type` parameter will accept an array of types to allow filtering.
+
+With support of the generic `PSResource`, this means we can also find and
+install arbitrary nupkgs that may only contain an assembly the user wants to
+use in their script.
+
+If the local cache does not exist, the cmdlet will call `Update-PSResourceCache`
+first and then attempt a search.
+However, this cmdlet does not update the cache if one exists.
+
+### Installing resources
+
+`Install-PSResource` will only work for modules.
+A `-AllowPrerelease` switch allows installing prerelease versions.
+Other types will use `Save-PSResource` (see below).
+A `-Repository` parameter accepts a specific repository name.
+If `-Repository` is not specified and the same resource is found in multiple
+repositories, then the resource is automatically installed quietly from the
+trusted repository with the highest version matching the `-Version` parameter
+(if specified, otherwise newest non-prerelease version unless `-AllowPrerelease`
+is used).
+If there are no trusted repositories matching the query, then the newest version
+fulfilling the query will be prompted to be installed.
+If there are multiple repositories with the same trust level containing the same
+version, the first one is used.
+
+`-AllowUntrusted` can be used to suppress being prompted for untrusted sources.
+`-AllowDifferentPublisher` can be used to suppress being prompted if the publisher
+of the module is different from the currently installed version.
+`-AllowReinstall` can be used to re-install a module.
+
+A `-Quiet` switch will suppress progress information.
+
+### Dependencies and version management
+
+`Install-PSResource` can accept a path to a psd1 file (using `-RequiredModulesFile`)
+or a hashtable (using `-RequiredModules`) where the key is the module name and the
+value is either the required version specified using Nuget version range syntax or
+a hash table where `repository` is set to the URL of the repository and
+`version` contains the [Nuget version range syntax](https://docs.microsoft.com/en-us/nuget/reference/package-versioning#version-ranges-and-wildcards).
```

Comment:
Will we also support the different formats of version numbers (2 part, 3 part, 4 part) as we have a mix already so need to support that better. Nuget itself has moved to pretty much enforcing 3 part/semver and that's causing issues for non-psgallery nuget feeds, just retaining the same version as the manifest for the resulting package file would be great.

Reply:
I think we would have to retain 4 part versions for compatibility.

Reply:
Would it be possible for us to [support both](https://github.com/PowerShell/PSScriptAnalyzer/blob/development/PSCompatibilityCollector/Microsoft.PowerShell.CrossCompatibility/Common/PowerShellVersion.cs)?

Reply:
The Nuget.Versioning dll supports 3 or 4 digits plus semver, as does the in-box SemVer, I think?

Reply:
When this change is made, built-in `*-Module` commands should be updated to support  `semver` as well as `[System.Version]` (for backwards compatibility). If the built-in commands are updated that brings up a question about how will older versions of PowerShell handle that. There are issues with the current paradigm of trying to force semver into a system that doesn't support it. We get issues like [PowerShell/PowerShellGet#324](https://github.com/PowerShell/PowerShellGet/issues/324) and [PowerShell/PowerShellGet#204](https://github.com/PowerShell/PowerShellGet/issues/204).

What will be the minimum PS version supported for `PowerShellGet` v3? If it's less than PS 7 then it may be worth breaking out the built-in `*-Module` commands from PowerShell into `PowerShellGet` to add support for semver in older PS versions.

Reply:
I would love it if PowerShell supported the `-prerelease.1` part of semver in module version folder names. The way it is now, pre-releases of modules with assemblies in them are a problem and fail to update if the module is loaded (like PSReadLine).

Reply:
> may be worth breaking out the built-in *-Module commands

PowerShell is pretty coupled to those -- the module system is a core part of PowerShell. I have imagined those commands as just cmdlet endpoints for PowerShell APIs (i.e. PowerShell owns the module system, but the cmdlets ship separately and slot in on top) but (1) that's a bucket of work and would probably break stuff and (2) doesn't solve the problem that old PowerShell versions still have the module cmdlets baked in -- no matter what you install, the only way to import a module in PS v3 of v5 is to call `Microsoft.PowerShell.Core\Import-Module`

Reply:
I think it's ok to have semver folder names so I'll add that.  This means that during transition, you might have two copies of the same module (like PSReadLine) since the old one is in a non-semver folder name.  Not going to attempt cleanup.

Regarding semver vs 4 part version, for now, I'll say we intend to support both, but we'll have to see what technical challenges come up.